tgui/packages/tgui/interfaces/ModularShieldGen.tsx

```diff
@@ -0,0 +1,147 @@
+import { useBackend } from '../backend';
+import { Window } from '../layouts';
+import { Stack, Section, ProgressBar, Button, NumberInput } from '../components';
+
+type modularShieldGeneratorData = {
+  max_strength: number;
+  current_strength: number;
+  max_regeneration: number;
+  current_regeneration: number;
+  max_radius: number;
+  current_radius: number;
+  active: Boolean;
+  recovering: Boolean;
+  exterior_only: Boolean;
+};
+
+export const ModularShieldGen = (props, context) => {
+  const { topLevel } = props;
+  const { act, data } = useBackend<modularShieldGeneratorData>(context);
+  const {
+    max_strength,
+    max_regeneration,
+    current_regeneration,
+    max_radius,
+    current_radius,
+    current_strength,
+    active,
+    exterior_only,
+    recovering,
+  } = data;
+
+  return (
+    <Window title="Modular Shield Generator" width={600} height={400}>
+      <Window.Content scrollable>
+        <Stack vertical fill>
+          <Stack.Item>
+            <Section
+              title={'Shield Strength'}
+              color={recovering ? 'red' : 'white'}>
+              <ProgressBar
+                title="Shield Strength"
+                value={current_strength}
+                minValue={0}
+                maxValue={max_strength}
+                height={'30px'}
+                ranges={{
+                  'good': [max_strength * 0.75, max_strength],
+                  'average': [max_strength * 0.25, max_strength * 0.75],
+                  'bad': [0, max_strength * 0.25],
+                }}>
+                {current_strength}/{max_strength}
+              </ProgressBar>
+            </Section>
+            <Section
+              horizontal
+              fill
+              height={'95px'}
+              title={'Regeneration and Radius'}>
+              <Section height={'20px'}>
+                <ProgressBar
+                  height={'20px'}
```

Comment:
fill and height are conflicting terms. also, it may be unavoidable, but try using grow={xnumber} instead of defining the sizes in pixels. you'll save yourself some brain cells later.